crates/ty_server/src/session.rs

```diff
@@ -302,16 +334,27 @@ impl Session {
                             .context("Failed to convert default options to metadata")
                             .and_then(|metadata| ProjectDatabase::new(metadata, system))
                             .expect("Default configuration to be valid");
+                    let default_root = db_with_default_settings
+                        .project()
+                        .root(&db_with_default_settings)
+                        .to_path_buf();
 
-                    self.projects.insert(
-                        db_with_default_settings
-                            .project()
-                            .root(&db_with_default_settings)
-                            .to_path_buf(),
-                        db_with_default_settings,
-                    );
+                    (default_root, db_with_default_settings)
                 }
-            }
+            };
+
+            // Carry forward diagnostic state if any exists
+            let previous = self.projects.remove(&root);
+            let untracked = previous
+                .map(|state| state.untracked_files_with_pushed_diagnostics)
+                .unwrap_or_default();
+            self.projects.insert(
+                root,
+                ProjectState {
+                    db,
+                    untracked_files_with_pushed_diagnostics: untracked,
+                },
+            );
```

Comment:
I think we need to call into `publish_settings_diagnostics` here to ensure the server pushes the diagnostics on initiale load.